blocks/search.py

```diff
@@ -0,0 +1,261 @@
+        ----------
+        \*\*kwargs :
+            Arguments needed by sequence generator
+
+        Returns
+        -------
+        Generated sequences
+        """
+        if not self.compiled:
+            self.compile()
+
+
+class GreedySearch(Search):
+
+    def __init__(self, sequence_generator):
+        super(GreedySearch, self).__init__(sequence_generator)
+
+
+class BeamSearch(Search):
+    """Beam search
+
+    Parameters
+    ----------
+    beam_size : int
+        Size of beam
+    batch_size : int
+        Size of input batch
+    sequence_generator : sequence generator
+        Sequence generator brick
+
+    """
+    def __init__(self, beam_size, batch_size, sequence_generator, attended,
+                 y, attended_mask, y_mask, inputs_dict, inputs):
+        super(BeamSearch, self).__init__(sequence_generator)
+        self.beam_size = beam_size
+        self.batch_size = batch_size
+        self.sequence_generator = sequence_generator
+        self.attended = attended
+        self.y = y
+        self.attended_mask = attended_mask
+        self.y_mask = y_mask
+        self.inputs_dict = inputs_dict
+        self.inputs = inputs
+        self.generate_names = sequence_generator.generate_outputs()
+        self.init_computer = None
+        self.next_computer = None
+
+    def compile(self, *args, **kwargs):
+        """Compiles functions for beamsearch
+
+        Parameters
+        ----------
+        inputs : dict
+            Dictionary of named inputs
+
+        """
+        generator = self.sequence_generator
+        attended = self.attended
+        attended_mask = self.attended_mask
+        init_generated = generator.generate(attended=attended,
+                                            attended_mask=attended_mask,
+                                            iterate=False, n_steps=1,
+                                            batch_size=self.batch_size)
+        init_generated = OrderedDict(zip(self.generate_names, init_generated))
+        init_cg = ComputationGraph(init_generated.values())
+        init_readouts = VariableFilter(application=generator.readout.emit,
+                                       name='readouts')(init_cg.variables)[-1]
+        init_probs = generator.readout.emitter._probs(init_readouts)
+
+        self.init_computer = function(self.inputs,
+                                      init_generated.values() + [init_probs])
+
+        cur_variables = OrderedDict()
+        for name, value in init_generated.iteritems():
+            cur_value = tensor.zeros_like(value)
+            cur_value.name = name
+            cur_variables[name] = cur_value
+
+        input_names = (generator.state_names + generator.glimpse_names +
+                       ['outputs'])
+        generator_inputs = {name: val for name, val
+                            in cur_variables.iteritems()
+                            if name in input_names}
+        next_generated = generator.generate(attended=attended,
```

Comment:
I do not understand why you need `init_generate` and `next_generated`. The first, as you wrote it, will not give the initial states that you are probably willing to get this way.

For getting the initial states use the `initial_state` method of the `BaseRecurrent`. The names of the states can be accessed by `generate.states`. 

Off the top of my head I think your should need the following functions:

1) One to compute the initial states. I described above how you should be able to build this one.

2) One to compute probabilities for the next words given the current states. You are doing it right here: 
`generator.generate` builds these probabilities under the hood and we can simply retrieve them with the variable filter.

3) One that computes next states given the current states and the chosen outputs. For that you need a more advanced hack. You know that in the `generator.generate` computation graph the sampled outputs are annotated as the outputs of the `readout.emit` computation graph. So what you could to do is replace those with your input Theano variables. Use the power of annotations!

4) That should have been said first, but you also need to compute all the contexts your sequence generator needs, as you most likely don't want to do it multiple times.